File: 프로그래머스/1/159994. 카드 뭉치/카드 뭉치.cpp
#include <string>
#include <vector>

using namespace std;

string solution(vector<string> cards1, vector<string> cards2, vector<string> goal) {
    
    int idx1 = 0;
    int idx2 = 0;
    
    for (const auto& word : goal) {
        if (idx1 < cards1.size() && cards1[idx1] == word) {
            ++idx1;
        } else if (idx2 < cards2.size() && cards2[idx2] == word) {
            ++idx2;
        } else {
            return "No";
        }
    }
    
    return "Yes";
}

File: 프로그래머스/1/159994. 카드 뭉치/카드 뭉치.java
import java.util.ArrayDeque;
import java.util.Arrays;

class Solution {
    public String solution(String[] cards1, String[] cards2, String[] goal) {
        
        ArrayDeque<String> cardsDeque1 = new ArrayDeque<>(Arrays.asList(cards1));
        ArrayDeque<String> cardsDeque2 = new ArrayDeque<>(Arrays.asList(cards2));

        for (String target : goal) {
            if (!cardsDeque1.isEmpty() && cardsDeque1.peekFirst().equals(target)) {
                cardsDeque1.pollFirst();
            } else if (!cardsDeque2.isEmpty() && cardsDeque2.peekFirst().equals(target)) {
                cardsDeque2.pollFirst();
            } else {
                return "No";
            }
        }

        return "Yes";
    }
}
